framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -528,56 +579,152 @@ private AnnotationMirror createBooleanAnnotationMirror(List<Boolean> values) {
+            for (int j = i; j < (1 + i) * rights.size(); j++) {
+                if (resultValues.get(j)) {
+                    thenLeftVals.add(lefts.get(i));
+                } else {
+                    elseLeftVals.add(lefts.get(i));
+                }
+            }
+        }
+
+        List<Number> thenRightVals = new ArrayList<>();
+        List<Number> elseRightVals = new ArrayList<>();
+
+        // Similarly to the left list, we need to iterate in an interesting way to get the appropriate values out
+        // of resultValues. In particular, we need to skip over large portions of the array at each step (because
+        // all the booleans that are relevant to a given right value each have rights.size - 1 other elements
+        // between them.
+        for (int i = 0; i < rights.size(); i++) {
+            for (int j = i; j < resultValues.size(); j += rights.size()) {
+                if (resultValues.get(j)) {
+                    thenRightVals.add(rights.get(i));
+                } else {
+                    elseRightVals.add(rights.get(i));
+                }
+            }
+        }
+
+        FlowExpressions.Receiver leftRec =
+                FlowExpressions.internalReprOf(analysis.getTypeFactory(), leftNode);
+        TypeMirror leftUnderlyingType =
+                atypefactory.getAnnotatedType(leftNode.getTree()).getUnderlyingType();
+
+        FlowExpressions.Receiver rightRec =
+                FlowExpressions.internalReprOf(analysis.getTypeFactory(), rightNode);
+        TypeMirror rightUnderlyingType =
+                atypefactory.getAnnotatedType(rightNode.getTree()).getUnderlyingType();
+
+        CFStore thenStore = result.getThenStore();
+        CFStore elseStore = result.getElseStore();
+
+        createAnnotationFromResultsAndAddToStore(
+                thenStore, thenLeftVals, leftUnderlyingType, leftRec);
+        createAnnotationFromResultsAndAddToStore(
+                elseStore, elseLeftVals, leftUnderlyingType, leftRec);
+
+        createAnnotationFromResultsAndAddToStore(
+                thenStore, thenRightVals, rightUnderlyingType, rightRec);
+        createAnnotationFromResultsAndAddToStore(
+                elseStore, elseRightVals, rightUnderlyingType, rightRec);
+
         return resultValues;
     }
 
+    private void createAnnotationFromResultsAndAddToStore(
+            CFStore store,
+            List<?> results,
+            TypeMirror underlyingType,
+            FlowExpressions.Receiver rec) {
+        AnnotationMirror anno = atypefactory.createResultingAnnotation(underlyingType, results);
+        store.clearValue(rec);
+        store.insertValue(rec, anno);
+    }
+
     @Override
     public TransferResult<CFValue, CFStore> visitLessThan(
             LessThanNode n, TransferInput<CFValue, CFStore> p) {
         TransferResult<CFValue, CFStore> transferResult = super.visitLessThan(n, p);
+        ConditionalTransferResult<CFValue, CFStore> newResult =
```

Comment:
To be entirely honest I don't know why I seem to need to do this. But if I thread the existing transfer result (`transferResult`) through, then the refinements stop working. I'd appreciate it if you could shed some light on why that's happening.